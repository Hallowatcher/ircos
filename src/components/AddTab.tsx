
import * as React from 'react';

type Props = {
  clickAddTab: () => void
}

const divStyle: React.CSSProperties = {
  padding: '10px',
  margin: '0 2px',
  display: 'flex',
  height: 25,
  alignItems: 'center',
  cursor: 'pointer',
  backgroundColor: 'rgb(241, 196, 15)',
  fontWeight: 800,
  color: 'white',
  transform: 'skew(-15deg)',
  boxSizing: 'border-box',
  WebkitUserSelect: 'none'
}

export class AddTab extends React.Component<Props, any> {
  render() {
    return (
      <div style={divStyle} onClick={this.props.clickAddTab}>
        <span style={{ transform: 'skew(15deg)' }}>+</span>
      </div>
    )
  }
}