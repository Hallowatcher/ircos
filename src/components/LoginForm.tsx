import * as React from 'react';

type Props = {
  submitLogin: (user: string, pass: string) => void
}

type State = {
  user: string,
  pass: string
}

export class LoginForm extends React.Component<Props, State> {

  constructor(props: any) {
    super(props);

    this.state = {
      user: '',
      pass: ''
    }
  }

  inputChanged(event: any) {
    this.setState({
      [event.target.name]: event.target.value
    })
  }

  handleSubmit(event: any) {
    event.preventDefault();

    if (this.state.user !== '' && this.state.pass !== '')
      this.props.submitLogin(this.state.user, this.state.pass);
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit.bind(this)}>
        <input type="text" name="user" value={this.state.user} onChange={this.inputChanged.bind(this)} />
        <input type="password" name="pass" value={this.state.pass} onChange={this.inputChanged.bind(this)} />
        <input type="submit" />
      </form>
    )
  }
}