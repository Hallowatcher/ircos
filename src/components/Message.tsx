import * as React from 'react';

export enum UserType {
  regular,
  moderator,
  self
}

type Props = {
  user: string,
  message: string,
  sentDate: Date,
  userType?: UserType
}

const styles = {
  users: {
    regular: {
      color: '#ffc561',
      fontWeight: 800
    } as React.CSSProperties,
    moderator: {
      color: '#BF3434',
      fontWeight: 800
    } as React.CSSProperties,
    self: {
      color: '#CECECE',
      fontWeight: 800
    } as React.CSSProperties
  },
  message: {

  } as React.CSSProperties,
  time: {
    color: '#b2b2b2'
  } as React.CSSProperties
}

export class Message extends React.Component<Props, any> {
  render() {

    let userStyle = styles.users.regular;
    switch (this.props.userType) {
      case UserType.moderator:
        userStyle = styles.users.moderator
        break;
      case UserType.self:
        userStyle = styles.users.self
        break;
    }

    // Build date
    const hours = ('0' + this.props.sentDate.getHours()).slice(-2);
    const minutes = ('0' + this.props.sentDate.getMinutes()).slice(-2);

    return (
      <div>
        <span style={styles.time}>[{hours}:{minutes}]</span> <span style={userStyle}>{this.props.user}:</span> <span style={styles.message}>{this.props.message}</span>
      </div>
    )
  }
}