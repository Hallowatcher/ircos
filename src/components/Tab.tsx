
import * as React from 'react';

type Props = {
  tabName: string,
  tabClick: () => void,
  closeTab: () => void,
  isActive: boolean
}

const xStyle: React.CSSProperties = {
  padding: '0 10px',
  margin: '0 0 0 10px',
  backgroundColor: 'rgba(255, 255, 255, 0.15)',
  height: '100%',
  display: 'flex',
  alignItems: 'center',
  fontWeight: 400
}

const divStyle: React.CSSProperties = {
  padding: '0 0 0 10px',
  margin: '0 2px',
  display: 'flex',
  height: 25,
  alignItems: 'center',
  cursor: 'pointer',
  backgroundColor: 'rgba(255, 255, 255, 0.1)',
  fontWeight: 800,
  color: 'white',
  transform: 'skew(-15deg)',
  boxSizing: 'border-box',
  WebkitUserSelect: 'none'
}

const divStyleActive: React.CSSProperties = {
  ...divStyle,
  backgroundColor: 'rgba(255, 255, 255, 0.3)'
}

export class Tab extends React.Component<Props, null> {
  render() {
    return (
      <div style={this.props.isActive ? divStyleActive : divStyle}>
        <div onClick={this.props.tabClick.bind(this, this.props.tabName)} style={ { transform: 'skew(15deg)' } }>{this.props.tabName}</div>
        <div onClick={this.props.closeTab.bind(this, this.props.tabName)} style={xStyle}>&times;</div>
      </div>
    )
  }
}