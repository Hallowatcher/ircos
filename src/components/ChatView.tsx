
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Message, UserType } from './Message';
import * as moment from 'moment';

interface IProps {
  messages: any[];
  selfNick: string;
  userClicked: (user: string) => void;
  style?: React.CSSProperties;
}

export class ChatView extends React.Component<IProps, any> {

  private scrollAtBottom: boolean = true;
  private scrollAtTop: boolean = false;
  private topMessage: any = null;
  private historyChanged: boolean = false;
  private multiplier: number = 1;

  public componentWillUpdate(nextProps: IProps) {
    this.historyChanged = nextProps.messages.length !== this.props.messages.length;
    if (!this.historyChanged || !this.refs.container) { return; }

    const container = this.refs.container as HTMLDivElement;
    const scrollPos = container.scrollTop;
    const scrollBottom = container.scrollHeight - container.clientHeight;
    this.scrollAtBottom = (scrollBottom <= 0) || (scrollPos === scrollBottom);
  }

  public componentDidUpdate() {
    if (this.scrollAtBottom) {
      this.scrollToBottom();
    }
  }

  public render() {
    return (
      <div style={this.props.style} ref="container">
        {this.mapMessages()}
      </div>
    );
  }

  /* istanbul ignore next */
  private scrollToBottom() {
    const container = this.refs.container as HTMLDivElement;
    const scrollHeight = container.scrollHeight;
    const height = container.clientHeight;
    const maxScrollTop = scrollHeight - height;
    ReactDOM.findDOMNode(container).scrollTop = maxScrollTop > 0 ? maxScrollTop : 0;
  }

  private mapMessages() {
    return this.props.messages.map((msg: any, index: number) => {

        // User type check
        let userType = UserType.regular;
        if (msg.name === this.props.selfNick) {
          userType = UserType.self;
        }
        if (msg.name === 'System') {
          userType = UserType.system;
        }

        return (
          <Message
            key={index}
            message={msg.text}
            user={msg.name}
            userType={userType}
            sentDate={moment(msg.date)}
            userClicked={this.props.userClicked}
          />
        );
      });
  }
}
