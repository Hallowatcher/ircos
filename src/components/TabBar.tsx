
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import Tab from './Tab';
import { AddTab } from './AddTab';
import * as FlipMove from 'react-flip-move';

interface IProps {
  tabs: string[];
  currentChannel: string;

  tabClick: (name: string) => void;
  tabMove: (from: number, to: number) => void;
  tabClose: (name: string) => void;
  tabAdd: () => void;
}

const style: React.CSSProperties = {
  flexGrow: 1,
  padding: '0 5px',
  display: 'flex',
  overflowY: 'hidden',
  overflowX: 'auto',
  alignItems: 'flex-end'
};

export class TabBar extends React.Component<IProps, any> {

  public render() {
    const channelMap = this.props.tabs.map((channel, index) =>
    (
      <Tab
        key={channel}
        ref={channel}
        index={index}
        tabName={channel}
        tabClick={this.props.tabClick}
        tabMove={this.props.tabMove}
        closeTab={this.props.tabClose}
        isActive={this.props.currentChannel === channel}
      />
    )
    );

    return (
      <FlipMove style={style} duration={125} easing="ease-out">
        {channelMap}
        <AddTab clickAddTab={this.props.tabAdd} key="add" ref="add" />
      </FlipMove>
    );
  }
}
