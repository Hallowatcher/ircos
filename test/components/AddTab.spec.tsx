
import * as React from 'react';
import { AddTab } from '../../src/components/AddTab';
import * as enzyme from 'enzyme';

describe('AddTab', function() {

  let component: enzyme.ShallowWrapper<any, any>;
  let clickAddTab: any;
  beforeEach(function() {
    clickAddTab = jest.fn();
    component = enzyme.shallow(<AddTab clickAddTab={clickAddTab} />);
  });

  it('should render', function() {
    expect(component).toMatchSnapshot();
  });

  it('should have an onClick prop', function() {
    expect(component.props().onClick).toBeDefined();
  });

  it('should call prop on click', function() {
    component.simulate('click');
    expect(clickAddTab).toHaveBeenCalled();
  });
});
