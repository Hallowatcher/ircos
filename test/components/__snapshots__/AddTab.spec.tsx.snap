// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTab should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    onClick={[Function]}
    style={
        Object {
            "WebkitUserSelect": "none",
            "alignItems": "center",
            "backgroundColor": "rgb(241, 196, 15)",
            "boxSizing": "border-box",
            "color": "white",
            "cursor": "pointer",
            "display": "flex",
            "fontWeight": 800,
            "height": 25,
            "margin": "0 2px",
            "padding": "10px",
            "transform": "skew(-15deg)",
          }
    }
>
    <span
        style={
            Object {
                "transform": "skew(15deg)",
              }
        }
    >
        +
    </span>
</div>,
  "nodes": Array [
    <div
      onClick={[Function]}
      style={
            Object {
                  "WebkitUserSelect": "none",
                  "alignItems": "center",
                  "backgroundColor": "rgb(241, 196, 15)",
                  "boxSizing": "border-box",
                  "color": "white",
                  "cursor": "pointer",
                  "display": "flex",
                  "fontWeight": 800,
                  "height": 25,
                  "margin": "0 2px",
                  "padding": "10px",
                  "transform": "skew(-15deg)",
                }
      }
>
      <span
            style={
                  Object {
                        "transform": "skew(15deg)",
                      }
            }
      >
            +
      </span>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <AddTab
        clickAddTab={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": AddTab {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "clickAddTab": [Function],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          onClick={[Function]}
          style={
                    Object {
                              "WebkitUserSelect": "none",
                              "alignItems": "center",
                              "backgroundColor": "rgb(241, 196, 15)",
                              "boxSizing": "border-box",
                              "color": "white",
                              "cursor": "pointer",
                              "display": "flex",
                              "fontWeight": 800,
                              "height": 25,
                              "margin": "0 2px",
                              "padding": "10px",
                              "transform": "skew(-15deg)",
                            }
          }
>
          <span
                    style={
                              Object {
                                        "transform": "skew(15deg)",
                                      }
                    }
          >
                    +
          </span>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          onClick={[Function]}
          style={
                    Object {
                              "WebkitUserSelect": "none",
                              "alignItems": "center",
                              "backgroundColor": "rgb(241, 196, 15)",
                              "boxSizing": "border-box",
                              "color": "white",
                              "cursor": "pointer",
                              "display": "flex",
                              "fontWeight": 800,
                              "height": 25,
                              "margin": "0 2px",
                              "padding": "10px",
                              "transform": "skew(-15deg)",
                            }
          }
>
          <span
                    style={
                              Object {
                                        "transform": "skew(15deg)",
                                      }
                    }
          >
                    +
          </span>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <AddTab
    clickAddTab={[Function]}
/>,
}
`;
